component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -0,0 +1,202 @@
+                }
+                return false;
+            }
+
+            String tenantDomain = OAuth2Util.getTenantDomainOfOauthApp(oAuthClientAuthnContext.getClientId());
+            X509Certificate registeredCert = (X509Certificate) OAuth2Util
+                    .getX509CertOfOAuthApp(oAuthClientAuthnContext.getClientId(), tenantDomain);
+            return authenticate(registeredCert, requestCert);
+
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.SERVER_ERROR, "Error occurred while retrieving " +
+                    "public certificate of client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        } catch (InvalidOAuthClientException e) {
+            throw new OAuthClientAuthnException(OAuth2ErrorCodes.INVALID_CLIENT, "Error occurred while retrieving " +
+                    "tenant domain for the client ID: " + oAuthClientAuthnContext.getClientId(), e);
+        }
+    }
+
+    /**
+     * Returns whether the incoming request can be authenticated or not using the given inputs.
+     *
+     * @param request    HttpServletRequest which is the incoming request.
+     * @param bodyParams Body parameters present in the request.
+     * @param context    OAuth2 client authentication context.
+     * @return Whether client can be authenticated using this authenticator.
+     */
+    @Override
+    public boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams,
+                                   OAuthClientAuthnContext context) {
+
+        if (clientIdExistsAsParam(bodyParams) && validCertExistsAsAttribute(request)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Client ID exists in request body parameters and a valid certificate found in request " +
+                        "attributes. Hence returning true.");
+            }
+            return true;
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Client id is not available in body params or valid certificate not found in request " +
+                        "attributes. Hence returning false.");
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Retrieves the client ID which is extracted from incoming request.
+     *
+     * @param request                 HttpServletRequest.
+     * @param bodyParams              Body parameter map of the incoming request.
+     * @param oAuthClientAuthnContext OAuthClientAuthentication context.
+     * @return Client ID of the OAuth2 client.
+     * @throws OAuthClientAuthnException OAuth client authentication Exception.
+     */
+    @Override
+    public String getClientId(HttpServletRequest request, Map<String, List> bodyParams, OAuthClientAuthnContext
+            oAuthClientAuthnContext) throws OAuthClientAuthnException {
+
+        Map<String, String> stringContent = getBodyParameters(bodyParams);
```

Comment:
What if client id param is not present ? Isn't this going to add a null entry if so ?

Reply:
canAuthenticate method validates the existence of a client ID. the authenticate method will not be invoked if the canAuthenticate returns false

Reply:
I would say the method needs to be self contained. It should not rely on the execution of previous methods. Ie if someone captures an instance of this class and try to extract client id, it should work

Reply:
added validation to authenticate method